### PR TITLE
enabling scheduler updates for every batch in training workflow

### PR DESCRIPTION
**Pull request by ImmoNathanael**
for example for torch.optim.lr_scheduler.CyclicLR or torch.optim.lr_scheduler.OneCycleLR.

### CODE REVIEW

**Comment by ImmoNathanael** [comment]
the workflows folder was not part of this repo, but since we are supposed to make changes in eisen-ai/covid19-challenge repo for the COVID19-challenge I had to make a copy of the trainworkflow.py from core and make a pull request for the whole file just to make a minor change. I want to make at least one more change to trainworkflow.py, is there a better way of doing this?  Thanks!

**Comment by faustomilletari** [comment]
first of all, thank you for the pull request. 

Let's try to find a way this will actually work and maybe be even easier.

What are you trying to do right now? are you maybe trying to do scheduling of the learning rate (like reducing the learning rate as you train).

The thing is that the workflow gets instantiated with a fixed list of poarameters. even tho your python is perfecly valid, i am unsure it would actually work during training. 

That said. Please explain to me what is your idea and i will try to make it happen or give you suggestions.

**Comment by ImmoNathanael** [comment]
Thanks for the quick reply! The way I understood the training workflow it is called every epoch and internally loops through the batches, inside trainworkflow.py. to me it wasn't obvious how to implement lr_schedulers like torch.optim.lr_scheduler.CyclicLR and torch.optim.lr_scheduler.OneCycleLR where an update to the lr is required after every training batch. 

The way I understand it the class Training(GenericWorkflow) only has a 'run' method, but no 'run_batch' method. 

In the colab notebook on the toy data this change works well, not sure how it would be to implement in the JSON for the platform. Is there an option to instantiate a scheduler?

**Comment by eisen-ai** [comment]
Maybe it would be easier to derive a new optimiser from the optimizer you intend to use
And then you could change the learning rate there. 

Let me look at the documentation.

I do think that there must be an internal parameter that the optimiser uses to keep the lr (probably something like self._lr)

So it would be

Class MyOptim(torch.optim.SGD):
	def __init__(self, scheduling_strategy: str, *args, **kwargs):
		….
		if scheduling_strategy == ‘Linear’:
			….
		elif scheduling…
		
		self._n_batches = 0
	def backward():
		self._lr = scheduler(self.lr, self._n_batches)
		super(…)
		
		self._n_batches += 1


I am just improvising here. But it would be definitely easier to change the optimiser rather than the workflow (and the optimiser is something you can config in the json).


Fausto

**Comment by ImmoNathanael** [comment]
> workflow gets instantiated with a fixed list of poarameters.

 I think I did add all required  parameters (batch_schedule, scheduler) to the training workflow _init_.

**Comment by eisen-ai** [comment]
I do agree that your code would be 100% valid if you were to code your own script in pure python.

But when you create a config.json, you would not have (as far as I know) anywhere to put additional parameters for the workflow.

In other words, if you use config files (and the CLI) which is what we are doing in the challenge, you would not be able to instantantiate the workflow as you wish but, instead, the workflow expects exactly the original fixed parameters list

Of course, if you are using eisen from python on the toy data, this strategy would work without issues. But then again, you would not need it merged into this repo...


Fausto

**Comment by ImmoNathanael** [comment]
> In other words, if you use config files (and the CLI) which is what we are doing in the challenge, you would not be able to instantantiate the workflow as you wish but, instead, the workflow expects exactly the original fixed parameters list

Thanks for the clarification! I didn't realize this was fixed, but thought it might be an evolving parameters list. I'll try to implement it following your suggestion :)
